src/main/java/org/takes/rs/Body.java

```diff
@@ -0,0 +1,176 @@
+         * @param content The {@link URL} of the content.
+         */
+        URLContent(final URL content) {
+            this.url = content;
+        }
+
+        @Override
+        public InputStream input() throws IOException {
+            return this.url.openStream();
+        }
+
+        @Override
+        public int length() throws IOException {
+            try (final InputStream input = this.url.openStream()) {
+                return input.available();
+            }
+        }
+    }
+
+    /**
+     * Content of a body based on a byte array.
+     */
+    final class ByteArrayContent implements Body {
+
+        /**
+         * The content of the body in a byte array.
+         */
+        private final transient byte[] bytes;
+
+        /**
+         * Constructs an {@code ByteArrayContent} with the specified byte array.
+         * @param content The content of the body.
+         */
+        ByteArrayContent(final byte[] content) {
+            this.bytes = content.clone();
+        }
+
+        @Override
+        public InputStream input() {
+            return new ByteArrayInputStream(this.bytes);
+        }
+
+        @Override
+        public int length() {
+            return this.bytes.length;
+        }
+    }
+
+    /**
+     * The content of the body based on an {@link InputStream}.
+     */
+    final class Stream implements Body {
+
+        /**
+         * The temporary file that contains the content of the body.
+         */
+        private final transient File file;
+
+        /**
+         * Constructs an {@code Stream} with the specified
+         * {@link InputStream}.
+         * @param input The content of the body as stream.
+         */
+        Stream(final InputStream input) {
+            this.file = Body.Stream.content(input);
+        }
+
+        @Override
+        public InputStream input() throws FileNotFoundException {
+            return new FileInputStream(this.file);
+        }
+
+        @Override
+        public int length() {
+            return (int) this.file.length();
+        }
+
+        /**
+         * Gives the {@code File} that contains the content of the provided
+         * {@link InputStream}.
+         * @param input The stream to store into the file.
+         * @return The {@code File} in which we stored the content of the
+         *  stream.
+         */
+        private static File content(final InputStream input) {
+            final File file;
+            try {
+                file = File.createTempFile("InputStreamContent", "tmp");
+                file.deleteOnExit();
+                try (final InputStream content = input) {
+                    Files.copy(
+                        content,
+                        Paths.get(file.getAbsolutePath()),
+                        StandardCopyOption.REPLACE_EXISTING
+                    );
+                }
+            } catch (final IOException ex) {
```

Comment:
@yegor256 Actually the thread safety is not even the point here, the point here is more the fact that streams once closed are not supposed to be reused even less to be reused by concurrent threads.